sources/healthdog/src/healthdog.rs

```diff
@@ -0,0 +1,160 @@
+/// are running.
+pub(crate) struct Healthdog {
+    /// The `Healthdog` configuration, e.g. from `/etc/healthdog.toml`
+    config: Config,
+    /// Information about the Bottlerocket release, e.g. from `os-release`
+    os_release: BottlerocketRelease,
+    /// A trait object that checks if a service (listed in `config`) is healthy. This can be passed-
+    /// in, but defaults to an object that uses `systemctl` to check services.
+    healthcheck: Box<dyn ServiceCheck>,
+}
+
+impl Healthdog {
+    /// Create a new instance by optionally passing in the `Config`, `BottlerocketRelease`, and
+    /// `ServiceCheck` objects. For each of these, if `None` is passed, then the default is used.
+    pub(crate) fn from_parts(
+        config: Option<Config>,
+        os_release: Option<BottlerocketRelease>,
+        healthcheck: Option<Box<dyn ServiceCheck>>,
+    ) -> Result<Self> {
+        Ok(Self {
+            config: match config {
+                None => Config::new()?,
+                Some(c) => c,
+            },
+            os_release: match os_release {
+                None => BottlerocketRelease::new().context(error::BottlerocketRelease)?,
+                Some(b) => b,
+            },
+            healthcheck: healthcheck.unwrap_or_else(|| Box::new(SystemdCheck {})),
+        })
+    }
+
+    /// # Description
+    ///
+    /// Sends key-value pairs as query parameters in a GET request to the URL in `config`. A
+    /// standard set of key-value pairs are added first, and appended by any additional parameters
+    /// passed in to this function.
+    ///
+    /// # Parameters
+    ///
+    /// * `sender`:          This is the name of the application sending the metrics e.g.
+    ///                      `healthdog` or `updog`.
+    /// * `event`:           The name of the type of metrics event that is being sent. For example
+    ///                      `boot-success` or `health-ping`.
+    /// * `values`:          The key-value pairs that you want to send. These will be sorted by key
+    ///                      before sending to ensure consistency of key-value ordering.
+    /// * `timeout_seconds`: The timeout setting for the HTTP client. Defaults to
+    ///                      `DEFAULT_TIMEOUT_SECONDS` when `None` is passed.
+    pub(crate) fn send<S1, S2>(
+        &self,
+        sender: S1,
+        event: S2,
+        values: Option<&HashMap<String, String>>,
+        timeout_seconds: Option<u64>,
+    ) -> Result<()>
+    where
+        S1: AsRef<str>,
+        S2: AsRef<str>,
+    {
+        let mut url = Url::from_str(&self.config.metrics_url).context(error::UrlParse {
+            url: self.config.metrics_url.clone(),
+        })?;
+        {
+            let mut q = url.query_pairs_mut();
+            q.append_pair("sender", sender.as_ref());
+            q.append_pair("event", event.as_ref());
+            let version = format!("{}", &self.os_release.version_id);
+            q.append_pair("version", &version);
```

Comment:
```suggestion
            q.append_pair("version", self.os_release.version_id.to_string());
```
Unless this specifically needs to be a reference, but `to_string` is still simpler/clearer than `format`, I think.

Same below for ignore-waves.

Reply:
Fixed in [2a5aef0](https://github.com/bottlerocket-os/bottlerocket/compare/f515d383c9e181c8a531a02414b320ff9231a81f..2a5aef04d7abcae5c91bf8cec84b8e95c7ccd306)